### PR TITLE
feat(plugin): BetterScreenshare, BetterMicrophone, PhilsPluginLibrary

### PR DESCRIPTION
**Pull request by philhk**
# BetterScreenshare

### Features
- **Profileable**
- **Custom Framerate**
- **Custom Resolution**
- **Custom Keyframe Interval**
- **Custom Video Bitrate**
- **Custom Video Codec**
- **Custom Audio Settings (identical to those in BetterMicrophone)**
- **HDR**
- **Hide default screenshare settings**
- **Custom settings panel button**
- **Simple mode which only offers (Custom Framerate, Custom Video Bitrate (with a list of options) and Custom Resolution (with a list of options))**

image
image
image
image
image
image
image

# BetterMicrophone

### Features
- **Profileable**
- **Custom Channels (Stereo)**
- **Custom Audio Bitrate**
- **Custom Sample Frequency**
- **Custom Sample Rate**
- **Custom Pac Size**
- **Simple mode which only offers (Stereo toggle and Custom Audio Bitrate (with a list of options))**

image
image
image

Both plugins are independent of each other so that means both are seperate plugins. If you'd enabled both the settings panel would look like this

image

# PhilsPluginLibrary

This plugin makes it easy to add a button to the settings panel and also contains some utility functions and frequently used components.

### CODE REVIEW

File: src/plugins/betterScreenshare/utils/deepmerge.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    };
+
+    const result = merge(...objects);
+
+    merge.options = defaultOptions;
+
+    return result;
+};
+
+export { merge };
```
Comment (Vap0r1ze) [inline src/plugins/betterScreenshare/utils/deepmerge.ts:122]: Does this do anything that lodash's [merge](https://lodash.com/docs/4.17.15#merge) doesn't?

Comment (philhk) [reply src/plugins/betterScreenshare/utils/deepmerge.ts]: No it does not. I totally forgot to remove it.

**Comment by abUwUser** [comment]
may you say which features does it add on the description of this PR? it is good to do that so that you do not need to look to the source code to see which changes were made

**Comment by philhk** [comment]
Of course! I've added a [README](https://github.com/philhk/Vencord/tree/main/src/plugins/betterScreenshare) explaining how each feature works, but here's a list.
- **Custom Screenshare Audio Source no matter what you are streaming**
- **Custom Screenshare Resolution**
- **Custom Screenshare Framerate**
- **Custom Screenshare Keyframe Interval**
- **Custom Screenshare Video Bitrate**
- **Custom Screenshare Audio Bitrate**
- **Custom Screenshare Video Codec**
- **Screenshare HDR Toggle**
- **Screenshare Profiles (they store the above mentioned settings)**
- **A toggle to hide the default screen sharing settings in the plugin options**
- **Everything is toggleable so no default features are removed from discord**

**Comment by abUwUser** [comment]
i meant to list them in the PR's __description__, not in a comment. it is easier to people to check it on the PR's description than in the readme or in a comment, again to avoid having to see the source code

**Comment by philhk** [comment]
Oh Okay I updated the description and added some pictures.

**Comment by abUwUser** [comment]
> Oh Okay I updated the description and added some pictures.

thanks. i appreciate your effort

**Comment by philhk** [comment]
Okay I've gone through all the files and checked everything and think it's done. There are probably things that can be improved but I can't see anything from my perspective but maybe someone with more experience.

**Comment by ThaUnknown** [comment]
would be really cool to see this on web too

the main issue being that web doesn't use this UI, instead skips it and goes into `navigator.mediaDevices.getDisplayMedia` and calls it with hardcoded 720p 30fps, regardless of what the video quality manager specifies so you'd need to do something along the lines of:
```js
const gdm = navigator.mediaDevices.getDisplayMedia

navigator.mediaDevices.getDisplayMedia = () => {
  return gdm.call(navigator.mediaDevices, yourQualityOptions)
}
```
or use reflect if you dislike call
getDisplayMedia parameters: https://devdocs.io/dom/mediadevices/getusermedia#parameters
so you'd preferably do something like
```
{
  frameRate: 120,
  video: {
    height: 1080,
    width: 1920
  },
  audio: true
}
```

I think the best way to accomplish this on web would be to patch `getDesktopSource` detect if its web, then force display your custom quality options UI, save its options, await user closing that ui, return to the original getDesktopSource call end, then pass those quality options inside GDM

I'm uncertain about how codecs are handled on web tho, so I can't help there, but it could probably be skipped because just being able to set quality and frameRate would be amazing

**Comment by exitss** [comment]
When I use this plugin and stream, Discord randomly restarts.

**Comment by exitss** [comment]
> When I use this plugin and stream, Discord randomly restarts.

It appears that the issue only occurs when the framerate setting is set to a high value.

**Comment by philhk** [comment]
I hope it's okay that I added my own plugin library. I already asked on the Discord but no one could tell me if it is. Maybe someone here can tell me :).

**Comment by ArjixWasTaken** [comment]
image


Good luck having ven review this

**Comment by philhk** [comment]
If I'm honest I'm glad at the moment because I found a bug I was trying to fix but after hours and hours of searching for the cause I gave up and now I think I'll take a break from this plugin. And just in case someone using my fork reads this, I'll explain the bug very quickly. It's pretty simple, if you have your camera on and end a screen share, the screen share settings will override the camera settings.

**Comment by ArjixWasTaken** [comment]
I'd suggest splitting this into multiple smaller PRs.
That way it will be easier to review and will be merged faster.

Bugs will be easier to figure out as well.

**Comment by somerandomcloud** [comment]
Honestly, since plugins are built in, there shouldn't be any need to add author-specific libraries. Each function should be made available inside the vencord API, if that makes sense.

**Comment by ArjixWasTaken** [comment]
> Honestly, since plugins are built in, there shouldn't be any need to add author-specific libraries. Each function should be made available inside the vencord API, if that makes sense. 

Unless said lib is huge enough to impact the bundle size.

**Comment by TheKodeToad** [comment]
> Honestly, since plugins are built in, there shouldn't be any need to add author-specific libraries. Each function should be made available inside the vencord API, if that makes sense.

I feel the same way. Maybe the api should be descriptive of what it is instead of being called Phil (I like the name though)!

**Comment by ArjixWasTaken** [comment]
Naming smth after the author is such a devilbro thing

**Comment by TheKodeToad** [comment]
This isn't really on-topic any more I don't think but it's quite a common thing with modding. I've seen it in Minecraft too lol!

**Comment by exitss** [comment]
closed :(

**Comment by iwa** [comment]
couldn't this get you banned anyways? they might have some warnings about bandwidth usage, kind of the same as twitch does, so even if it's a good idea, i don't think it'd be worth the risk...

**Comment by philhk** [comment]
I've been using this plugin daily since creating this PR at the max bitrate of 10000kbps and nothing has happened. You decide whether to take the risk.

**Comment by zxsilenx** [comment]
Can't build your vencord fork:

`Build failed with 3 errors:
src/plugins/betterMicrophone.desktop/logger/index.ts:19:7: ERROR: No matching export in "src/utils/Logger.ts" for import "default"
src/plugins/betterScreenshare.desktop/logger/index.ts:19:7: ERROR: No matching export in "src/utils/Logger.ts" for import "default"
src/plugins/philsPluginLibrary/discordModules/components.tsx:19:9: ERROR: No matching export in "src/utils/misc.tsx" for import "LazyComponent"
2 warnings and 3 errors
 ELIFECYCLE  Command failed with exit code 1.
`

**Comment by ArjixWasTaken** [comment]
> Can't build your vencord fork:
> 
> `Build failed with 3 errors: src/plugins/betterMicrophone.desktop/logger/index.ts:19:7: ERROR: No matching export in "src/utils/Logger.ts" for import "default" src/plugins/betterScreenshare.desktop/logger/index.ts:19:7: ERROR: No matching export in "src/utils/Logger.ts" for import "default" src/plugins/philsPluginLibrary/discordModules/components.tsx:19:9: ERROR: No matching export in "src/utils/misc.tsx" for import "LazyComponent" 2 warnings and 3 errors  ELIFECYCLE  Command failed with exit code 1. `

I can build it just fine, so you are doing smth wrong.

**Comment by ArjixWasTaken** [comment]
Dunno if this is already a known issue, but on web (and vencord desktop) this plugin makes it impossible to connect to a VC.
It is stuck at "RTC Connecting".

**Comment by philhk** [comment]
That's weird, it probably has something to do with something else because I'm not having any issues at all. Have you tried restarting your PC?

**Comment by ArjixWasTaken** [comment]
did you try on normal discord?
or did you mean vencord desktop? vencord desktop is discord web on an electron wrapper

**Comment by zxsilenx** [comment]
it's fine, we solved this with dev

**Comment by Loukious** [comment]
Why was this closed btw?

**Comment by 3941** [comment]
No idea, but I'd REALLY like to see this plugin be a thing.
If anyone has some pointers towards how to use this manually, I'd be very happy.
Just modifying the video bitrate would be enough for me, actually.

**Comment by 3941** [comment]
Thank you so much @EyeDeck !
I did run into an ExecutionPolicy issue, but after setting it to unrestricted temporarily it all worked just fine!

**Comment by berksmbl** [comment]
I think the development was stopped before it could be released :(
Sadly, it was a promising project

**Comment by 3941** [comment]
It works fine as it is. I think the problem is rather the review process. Since Vencord has every plugin built in, and no way to add them manually, everything needs to be reviewed thoroughly before it gets added to the project. And this is a HUGE plugin(s) to review.

Not sure why it was closed. Maybe they didn't like something. Maybe they didn't want to bother reviewing it.
But it works. I'm using philhk's fork now. Ofc that means I won't get any other updates, but right now it does everything I need it to do and it runs very stable.

**Comment by berksmbl** [comment]
@3941 I don't want to break away from the official version.

**Comment by lewisakura** [comment]
> Not sure why it was closed. Maybe they didn't like something. Maybe they didn't want to bother reviewing it.

The author themselves closed the PR. I am unsure if this is due to discussions in Discord or whatnot.

From a contributor standpoint, this is a ***HUGE*** change and as aforementioned would be better suited for when the third-party plugin system comes out fully. The only reasonable way this would get reviewed would be if this was split into two separate PRs (one for the BetterScreenshare plugin, one for the BetterMicrophone plugin).

On top of this, however, there are a few issues with this PR:
- BetterMicrophone violates Vencord's "no stereo mic" rule, though I suppose this is more of a thing to stop people from asking for it since we don't particularly want to support it.
- Plugin libraries aren't really a thing in Vencord and the abilities this library has should be turned into a general API plugin or into part of Vencord core.
- The way the plugins are written is highly unusual (using a class?) and will potentially break in the future since it is not using `definePlugin`. This is a relatively easy fix but I am unsure why they are written this way.
- The plugins have direct references to the library plugin's files in its imports (see point two).
- These plugins are absolutely massive compared to anything else in Vencord, which increases review time. There are roughly 100 different source code files we have to look through here compared to the 1-3 on average for new plugins (though from quick inspection a lot of these could be merged together).
- There are concerns that it is too risky. Vencord is meant to have guarantees that plugins are safe to use and enabling all of them shouldn't lead to issues. However, both of these plugins give control to parameters that allow to produce results very far outside the bounds of the regular client, which may lead to your account being at risk in the future. 
- I am unsure if this will continue to be maintained.

Again, you can run the fork directly, but we will not give support for doing so. Sorry.

**Comment by twisty-nado** [comment]
> Stupid question but how do I download this masterpiece?

also curious lol

**Comment by twisty-nado** [comment]
found it https://github.com/philhk/better-discord-screenshare-plugin

**Comment by 3941** [comment]
> Stupid question but how do I download this masterpiece?

No point anymore. The branch was deleted and afaik never developed any further, while Discord changed many things in the meantime. Best you can do, for better screenshare anyways, is Betterdiscord and YABDP4Nitro plugin.

@twisty-nado that's been deprecated even before the Vencord version was started. Useless, I'm afraid.

**Pull request closed without merging**